service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
+            }
+        }
+    }
+
+    void LevelZeroImp::metric_destroy(unsigned int l0_device_idx, unsigned int l0_domain_idx)
+    {
+////        //std::cout << "ZET_DESTROY - gpu " << std::to_string(l0_device_idx) << std::endl;
+//        ze_result_t ze_result;
+//
+//        if (m_devices.at(l0_device_idx).metrics_initialized) {
+//            m_devices.at(l0_device_idx).metrics_initialized = false;
+//
+//            // Close metric streamer
+//            ze_result = zetMetricStreamerClose(m_devices.at(l0_device_idx).metric_streamer);
+//            check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                            "LevelZero::" + std::string(__func__) +
+//                            ": LevelZero Metric Streamer Close failed",
+//                            __LINE__);
+//
+//            // Destroy Event
+//            ze_result = zeEventDestroy(m_devices.at(l0_device_idx).event);
+//            check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                            "LevelZero::" + std::string(__func__) +
+//                            ": LevelZero Metric Event Destroy failed",
+//                            __LINE__);
+//
+//            // Destroy Event Pool
+//            ze_result = zeEventPoolDestroy(m_devices.at(l0_device_idx).event_pool);
+//            check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                            "LevelZero::" + std::string(__func__) +
+//                            ": LevelZero Metric Event Pool Destroy failed",
+//                            __LINE__);
+//
+//            // Deactivate the device context
+//            ze_context_handle_t context = m_devices.at(l0_device_idx).context;
+//            ze_result = zetContextActivateMetricGroups(context,
+//                                                       m_devices.at(l0_device_idx).device_handle,
+//                                                       0, nullptr);
+//            check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+//                            "LevelZero::" + std::string(__func__) +
+//                            ": LevelZero Metric Context Deactivation failed",
+//                            __LINE__);
+//        }
+    }
+
+
+    void LevelZeroImp::metric_init(unsigned int l0_device_idx, unsigned int l0_domain_idx)
+    {
+        ze_result_t ze_result;
+        ze_context_handle_t context = m_devices.at(l0_device_idx).subdevice.context.at(l0_domain_idx);
+        ze_result = zetContextActivateMetricGroups(context, m_devices.at(l0_device_idx).subdevice_handle.at(l0_domain_idx),
+                                                   1, &m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx));
+
+        ze_event_pool_handle_t event_pool_handle = nullptr;
+        ze_event_pool_desc_t event_pool_desc = {ZE_STRUCTURE_TYPE_EVENT_POOL_DESC, nullptr, 0, 1};
+
+        ze_result = zeEventPoolCreate(context, &event_pool_desc, 1,
```

Comment:
Knowledge check for myself. Docs say 1 here is the number of devices. Is this 1 because we're initializing one GPU at a time?

Reply:
From reading further down I guess it also means we're sampling one device at a time, since this is the event that read attempts block on while waiting for metric data

Reply:
We're targeting 1 event in each event pool, and 1 sub-device per event.